File: Problems/No.535_Encode_and_Decode_TinyURL/encode_and_decode_tinyurl.py
class Codec:

    def __init__(self):
        self.T2L = {}
        self.L2T = {}
        self.tinyURL_idx = 1

    def encode(self, longUrl):
        """Encodes a URL to a shortened URL.
        
        :type longUrl: str
        :rtype: str
        """
        if longUrl in self.L2T:
            return self.L2T[longUrl]
        new_tinyURL = "http://tinyurl.com/" + str(self.tinyURL_idx)
        self.tinyURL_idx += 1
        self.L2T[longUrl] = new_tinyURL
        self.T2L[new_tinyURL] = longUrl
    
        return new_tinyURL
        
        

    def decode(self, shortUrl):
        """Decodes a shortened URL to its original URL.
        
        :type shortUrl: str
        :rtype: str
        """
        return self.T2L[shortUrl]
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))

File: Problems/No.535_Encode_and_Decode_TinyURL/encode_and_decode_tinyurl.cpp
class Solution {

private:
    std::unordered_map<std::string, std::string> T2L;
    std::unordered_map<std::string, std::string> L2T;
    int tinyURL_idx = 1;

public:

    // Encodes a URL to a shortened URL.
    string encode(string longUrl) {

        if (L2T.count(longUrl)) return L2T[longUrl];
        std::string new_tinyURL = "http://tinyurl.com/" + std::to_string(tinyURL_idx);
        tinyURL_idx++;
        L2T[longUrl] = new_tinyURL;
        T2L[new_tinyURL] = longUrl;

        return new_tinyURL;
        
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl) {return T2L[shortUrl];}

};

// Your Solution object will be instantiated and called as such:
// Solution solution;
// solution.decode(solution.encode(url));
